hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/opt/physical/SortPhysicalRule.java

```diff
@@ -0,0 +1,186 @@
+
+            RelNode rel;
+
+            if (requiresLocalSort) {
+                rel = createLocalSort(logicalSort, physicalInput);
+            } else {
+                rel = createLocalNoSort(physicalInput);
+            }
+
+            // Add merge phase if needed.
+            DistributionTrait physicalInputDist = OptUtils.getDistribution(physicalInput);
+
+            if (!physicalInputDist.isFullResultSetOnAllParticipants()) {
+                rel = createMerge(rel, logicalSort);
+            }
+
+            // Add to the list of transformations.
+            if (requiresLocalSort) {
+                requiresLocalSortRels.add(rel);
+            } else {
+                noLocalSortRels.add(rel);
+            }
+        }
+
+        // Pick up a plan that doesn't require local sorting,
+        // otherwise an exception will be thrown when local sort
+        // is touched in runtime.
+        if (noLocalSortRels.size() > 0) {
+            return noLocalSortRels;
+        } else {
+            return requiresLocalSortRels;
+        }
+    }
+
+    /**
+     * Check if local sorting phase is needed. It could be avoided iff sort collation is a prefix of input collation.
+     *
+     * @return {@code true} if local sorting is needed, {@code false} otherwise.
+     */
+    private static boolean requiresLocalSort(RelCollation sortCollation, RelCollation inputCollation) {
+        if (sortCollation.getFieldCollations().isEmpty()) {
+            // No need for sorting
+            return false;
+        }
+
+        List<RelFieldCollation> sortFields = sortCollation.getFieldCollations();
+        List<RelFieldCollation> inputFields = inputCollation.getFieldCollations();
+
+        if (sortFields.size() <= inputFields.size()) {
+            for (int i = 0; i < sortFields.size(); i++) {
+                RelFieldCollation sortField = sortFields.get(i);
+                RelFieldCollation inputField = inputFields.get(i);
+
+                // Different collation, local sorting is needed.
+                if (!sortField.equals(inputField)) {
+                    return true;
+                }
+            }
+
+            // Prefix is confirmed, no local sorting is needed.
+            return false;
+        } else {
+            // Input has less collated fields than sort. Definitely not a prefix => local sorting is needed.
+            return true;
+        }
+    }
+
+    private static SortPhysicalRel createLocalSort(SortLogicalRel logicalSort, RelNode physicalInput) {
+        // Input traits are propagated, but new collation is used.
+        RelTraitSet traitSet = OptUtils.traitPlus(physicalInput.getTraitSet(),
+            logicalSort.getCollation()
+        );
+
+        return new SortPhysicalRel(
+            logicalSort.getCluster(),
+            traitSet,
+            physicalInput,
+            logicalSort.getCollation()
+        );
+    }
+
+    private RelNode createLocalNoSort(RelNode input) {
+        return input;
```

Comment:
Minor: probably we do not need this method at all.

Reply:
Addressed in https://github.com/hazelcast/hazelcast/pull/18013/commits/f9419cfe53b49156bfacd109deae81e73631f1c5